### PR TITLE
pulling server config from a config file rather than command line flags

### PR DESCRIPTION
**Pull request by sminez**

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    /// Relative path to a valid private key in PEM format
+    pub private_key_path: PathBuf,
+    /// Activitypub related configuration for the relay
+    pub activity_pub: ActivityPubConfig,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            listen: Ipv4Addr::new(0, 0, 0, 0),
```
Comment (dmah42) [inline src/config.rs:27]: the default should be safe: 127, 0, 0, 1

File: src/config.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            port: 4242,
+            data_dir: PathBuf::from("."),
+            private_key_path: PathBuf::from("private-key.pem"),
```
Comment (dmah42) [inline src/config.rs:30]: do we need a default for the key? it's always required right?

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+pub mod routes;
+pub mod signature;
+pub mod state;
+pub mod util;
+
+pub use error::{Error, Result};
+
+/// Lookup our base url from the environment or default to localhost:4242
+pub fn base_url() -> &'static str {
+    // TODO: move this to Args or build it from there.
```
Comment (dmah42) [inline src/lib.rs:13]: can we fix this now we have the config?

File: src/main.rs [hunk 4]
```diff
@@ -1,45 +1,31 @@
...
-pub use error::{Error, Result};
+use actiserve::{
+    base_url,
+    config::Config,
+    routes::build_routes,
+    state::{Db, State},
+};
 
 // TODO: move this to Args
 const PORT: u16 = 4242;
```
Comment (dmah42) [inline src/main.rs:15]: do we still need this?

File: src/config.rs (hunk 2 above)
Comment (sminez) [reply src/config.rs:30]: The defaults are here for being able to write out the initial config file if there isn't one already. The alternative is to just include an example one I suppose?

File: src/lib.rs (hunk 3 above)
Comment (sminez) [reply src/lib.rs:13]: ah, yeah we can 👍

File: src/main.rs (hunk 4 above)
Comment (sminez) [reply src/main.rs:15]: nope! The should be getting pulled from the config as well

File: resources/config.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# IPv4 address to listen on
```
Comment (dmah42) [inline resources/config.yaml:1]: should this be `resources/config.example.yml` to indicate that it should be changed by the end user?

File: src/config.rs (hunk 2 above)
Comment (dmah42) [reply src/config.rs:30]: yeah i assumed the resources/config.yml was the example one :)

File: resources/config.yaml (hunk 5 above)
Comment (sminez) [reply resources/config.yaml:1]: yeah that's probably worthwhile 🙂

File: src/config.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    /// Instances that should always be rejected
+    pub blocked_instances: Vec<String>,
+    /// Whether or not the allow list should be enabled (blocking
+    /// anything not on the list)
+    pub allow_list: bool,
+    /// Instances that should accepted. Only enforced if allowList=true
+    pub allowed_instances: Vec<String>,
+}
+
+impl Default for ActivityPubConfig {
```
[the remark this answers is not in the document]
Comment (sminez) [reply src/config.rs:55]: Arguably not: it's used in `new_with_test_key` in state.rs and being able to make default instances is (in general) useful, but given that that's the only place where we need it at the moment it's probably simpler to just inline this where it's used instead.

**Review by dmah42 — APPROVED** [review]

**Review by dmah42 — APPROVED** [review]

**Pull request merged**